### PR TITLE
Add example for hyperparameter tuning

### LINKED ISSUE #76: `aux_posterior` is not AD-ready
I tried to AD `aug_elbo` in the `NegBinomialLikelihood` example, i.e. (removed unnecessary bits), purposefully avoiding ParameterHandling.jl and trying only with `ForwardDiff.gradient`

```julia
# # Negative Binomial

# We load all the necessary packages
using AbstractGPs
using ApproximateGPs
using AugmentedGPLikelihoods
using Distributions
using ForwardDiff # <-- try this first
using LinearAlgebra

# We create some random data (sorted for plotting reasons)
N = 100
x = range(-10, 10; length=N)
kernel = with_lengthscale(SqExponentialKernel(), 2.0)
gp = GP(kernel)
lik = NegBinomialLikelihood(15)
lf = LatentGP(gp, lik, 1e-6)
f, y = rand(lf(x));

# ## ELBO
# How can one compute the Augmented ELBO?
# Again AugmentedGPLikelihoods provides helper functions
# to not have to compute everything yourself
function aug_elbo(lik, u_post, x, y)
    qf = marginals(u_post(x))
    qΩ = aux_posterior(lik, y, qf)
    return expected_logtilt(lik, qΩ, y, qf) - aux_kldivergence(lik, qΩ, y) -
           kldivergence(u_post.approx.q, u_post.approx.fz)     # approx.fz is the prior and approx.q is the posterior 
end

function u_posterior(fz, m, S)
    return posterior(SparseVariationalApproximation(Centered(), fz, MvNormal(m, S)))
end

# ## Try to differentiate loss function

function makeloss(x, y)
    N = length(x)
    function loss(θ)
        k = ScaledKernel(
            RBFKernel() ∘ ScaleTransform(inv(θ.lengthscale)), 
            θ.variance
        )
        gp = GP(k)
        lik = NegBinomialLikelihood(θ.r)
        fz = gp(x, 1e-8);
        u_post = u_posterior(fz, zeros(N), Matrix{Float64}(I(N)))
        return aug_elbo(lik, u_post, x, y)
    end
end

θ = (
    variance = 1.,
    lengthscale = 1.,
    r = 15.
)

θ_flat = vcat(
    θ.variance, 
    θ.lengthscale,
    θ.r
)

function vec_to_nt(v)
    return (
        variance = v[1],
        lengthscale = v[2],
        r = v[3]
    )
end

vec_to_nt(θ_flat) == θ

loss = makeloss(x, y)
loss(θ) # works!
ForwardDiff.gradient(loss ∘ vec_to_nt, θ_flat) # MethodError
```

There is an easy fix (happy to open a PR): change the definition of `aux_posterior` as
```julia
function aux_posterior(lik::AbstractLikelihood, y, f)
    v = sqrt.(second_moment.(f))
    return For(TupleVector(; y=y, c=v)) do φ
        NTDist(PolyaGamma(φ.y + lik.r, φ.c)) # Distributions uses a different parametrization
    end
    return aux_posterior!(init_aux_posterior(lik, length(y)), lik, y, f)
end
```

```julia
julia> ForwardDiff.gradient(loss ∘ vec_to_nt, θ_flat)
3-element Vector{Float64}:
  5.790557942012172e7
 -1.9761748845444782e9
 16.184871970106013
```

BTW: is it expected that the values of the augmented ELBO are so much larger in magnitude than the normal ELBO?

### PR DESCRIPTION
**Pull request by simsurace**
This is supposed to close #76

### CODE REVIEW

**Comment by theogf** [comment]
Nice thanks! If that's okay with you, could we merge #70 first and adapt this PR accordingly?

**Comment by simsurace** [comment]
I will be making the necessary adaptations shortly.

**Comment by simsurace** [comment]
The figure is probably not needed, but will be auto-generated by Literate, right?

File: examples/hyperpars_nb/script.jl [hunk 1]
```diff
@@ -0,0 +1,344 @@
...
+        if log(ac) - logvar < -0.5
+            i0 = i
+            break
+        end
+    end
+    return i0
+end
+
+# ## Diagonal SVGP model
+# We define a standard diagonal sparse variational GP for comparison
```
Comment (theogf) [inline examples/hyperpars_nb/script.jl:92]: I don't understand this. What is the point of having a diagonal svgp?

File: examples/hyperpars_nb/script.jl [hunk 2]
```diff
@@ -0,0 +1,344 @@
...
+    post = u_posterior(fz, l.m, l.S)
+    posty = LatentGP(post, lik, 1e-6)
+    residuals = l.y - mean(posty(l.x))
+    return posty, residuals
+end
+
+
+# ## Overloads for posterior over Y
+# We can calculate the mean and variance of Y analytically
+function Statistics.mean(f::AbstractGPs.LatentFiniteGP{T1, NegativeBinomialLikelihood{NBParamFailure{T2}, LogisticLink}}) where {T1, T2}
```
Comment (theogf) [inline examples/hyperpars_nb/script.jl:242]:
```suggestion
function Statistics.mean(f::AbstractGPs.LatentFiniteGP{T1, <:NegativeBinomialLikelihood{<:NBParamFailure, LogisticLink}}) where {T1}
```

File: examples/hyperpars_nb/script.jl [hunk 3]
```diff
@@ -0,0 +1,344 @@
...
+    return @. my * (1 + my * ((1 + inv(r)) * exp(c) - 1))
+end
+
+# We approximate the posterior over Y by a truncated Gaussian distribution
+function AbstractGPs.marginals(
+    f::AbstractGPs.LatentFiniteGP{T1, NegativeBinomialLikelihood{NBParamFailure{T2}, LogisticLink}}
+) where {T1, T2}
+    m, c = mean(f), var(f)
+    return @. truncated(Normal(m, sqrt(c)), zero(eltype(m)), nothing)
+end
```
Comment (theogf) [inline examples/hyperpars_nb/script.jl:261]: Why would you need this? The marginals should not be truncated?

File: examples/hyperpars_nb/script.jl [hunk 4]
```diff
@@ -0,0 +1,344 @@
...
+
+# We call and time the `optimize` function to find optimal parameters
+# The augmented GP leads to a considerable speedup, at least for this
+# problem size.
+@time θ1 = optimize(loss1)
+@time θ2 = optimize(loss2)
+
+# We evaluate the loss function after training
+@show loss1(θ1)
+@show loss2(θ2)
```
Comment (theogf) [inline examples/hyperpars_nb/script.jl:296]: Instead of `@show` you should use the Literate API 
```suggestion
loss1(θ1), loss2(θ2)
```

File: examples/hyperpars_nb/script.jl [hunk 5]
```diff
@@ -0,0 +1,344 @@
...
+
+# We construct the posterior and residuals after training
+post1, res1 = build_posterior(loss1, θ1)
+post2, res2 = build_posterior(loss2, θ2)
+
+# We calculate the mean squared errors
+@show mean(abs2, res1) / var(y)
+@show mean(abs2, res2) / var(y)
```
Comment (theogf) [inline examples/hyperpars_nb/script.jl:304]:
```suggestion
mean(abs2, res1) / var(y), mean(abs2, res2) / var(y)
```

**Review by theogf — COMMENTED** [review]
My main issue is with DSVGP, I don't understand why it needs to be introduced.
It would be more interesting to see what is the difference between the model with and without hyperparameters optimized.
Also I would rather have the `NonCentered` appproach, it looks like it is slowly getting solved upstream!

File: examples/hyperpars_nb/script.jl [hunk 6]
```diff
@@ -0,0 +1,344 @@
...
+    μ, σ² = initial_μ_σ²_Poisson(mean(l.y), var(l.y))
+    xmin = minimum(l.x)
+    xmax = maximum(l.x)
+    Δt = (xmax - xmin) / length(l.x)
+    invl0 = inv(correlation_length(l.y) * Δt)
+    r0 = 100.
+    M = min(30, length(l.y))
+    z0 = rand(M)
+    z0 .*= xmax - xmin
+    z0 .+= xmin
```
[the remark this answers is not in the document]
Comment (simsurace) [reply examples/hyperpars_nb/script.jl:109]: The reason why I'm interested in the augmented GP framework is to make training faster. The sparse variational GP is already much slower for moderate data sizes. If we make the variational GP use all training points, it would be hopelessly slow, wouldn't it?

File: examples/hyperpars_nb/script.jl [hunk 7]
```diff
@@ -0,0 +1,344 @@
...
+    fz = f(l.x).fx
+    N = length(l.x)
+    qΩ = init_aux_posterior(lik, N)
+    Zygote.@ignore cavi!(fz, lik, l.x, l.y, l.m, l.S, qΩ; niter=1)
+    L = aug_elbo(lik, u_posterior(fz, l.m, l.S), l.x, l.y)
+    return -L
+end
+
+# In order to make this work with Zygote, we had to change from
+# `Centered` to `NonCentered` here. This deviates from the usual practice.
```
[the remark this answers is not in the document]
Comment (simsurace) [reply examples/hyperpars_nb/script.jl:193]: Sure, but if we merge this before those issues get resolved, we have to have a workaround. 
Are there any concerns that the `NonCentered` representation is incorrect or less efficient?

File: examples/hyperpars_nb/script.jl (hunk 3 above)
Comment (simsurace) [reply examples/hyperpars_nb/script.jl]: This is a very rough approximation and makes sure that credible intervals are always nonnegative. This is now removed in favor of Monte Carlo estimates.

**Comment by simsurace** [comment]
Unfortunately after more testing it is still not working reliably. The convergence is sometimes very fast but the solution not very good.
Playing around with the data size, ground truth parameters, and random seed should reveal these problems.
